MongoDB intro closes #698 (#722)
bob-crutchley/courseware: topics/mongodb/modules/intro/README.md

# Intro

<!--TOC_START-->
## Contents
- [Overview](#overview)
	- [What is it?](#what-is-it)
		- [Database Type](#database-type)
		- [Data Format](#data-format)
		- [Structure](#structure)
			- [Collections](#collections)
			- [Databases](#databases)
- [Why is it used?](#why-is-it-used)
- [When should you use it?](#when-should-you-use-it)
- [Tutorial](#tutorial)
	- [Installation](#installation)
		- [Windows](#windows)
		- [Linux](#linux)
			- [Ubuntu](#ubuntu)
- [Exercises](#exercises)

<!--TOC_END-->
## Overview

> ### What is it?

Reviewer: ```suggestion

```

I wouldn't bother with this, just make the other headers three hashes deep instead

> #### Database Type

Reviewer: ```suggestion
### Database Type
```

MongoDB is a **document store**.
Document stores are *non-relational* databases - this means that each entry in the database (each 'document') represents a whole object rather than having objects spread across different tables.

> #### Data Format

Reviewer: ```suggestion
### Data Format
```

Documents in Mongo are stored in BSON (Binary JavaScript Object Notation) which, like JavaScript Object Notation, is a system that involves storing data in key-value pairs.
For example, a person could be represented in BSON as:

```javascript
{
    "_id" : ObjectId("5e135963c3782165689e6142"), 
    "firstName": "Tadas",

>     "secondName": "Vlaidotas",

Reviewer: ```suggestion
    "secondName": "Vaidotas",
```

>     "secondName": "Vlaidotas",

Reviewer: ```suggestion
    "secondName": "Vaidotas",
```

    "occupation": "Trainer",
    "age": 33,
    "specialisation": "DevOps" 
}
```

#### Structure

##### Collections

> Documents are grouped into collections based on their subject;

Reviewer: ```suggestion
Documents are grouped into collections based on their subject.
```

For example, a 'people' collection might be used to store documents like the one above:
```javascript
{ 
    "_id" : ObjectId("5e135963c3782165689e6142"), 
    "firstName" : "Tadas", 

>     "lastName" : "Vlaidotas", 

Reviewer: ```suggestion
    "lastName" : "Vaidotas", 
```

    "age" : "33", 
    "occupation" : "Trainer",
    "specialisation": "DevOps" 
}
{ 
    "_id" : ObjectId("5e13599f9138596568d873ca"), 
    "firstName" : "Jordan", 
    "lastName" : "Harrison", 
    "age" : "25", 
    "occupation" : "Trainer" 
}
```
While documents in a collection should have *something* in common, Mongo places no restriction on their structure.

> Documents in a collection can have identical fields or basically nothing in common, it's entirely up to the person entering the data.

Reviewer: ```suggestion
Documents in a collection are flexible in composition - they might have identical fields, or basically nothing in common.

This leaves the composition of the data entered into that document entirely up to the person entering the data.
```

Notice how, in the example above, Tadas has a specialisation whereas Jordan does not.

##### Databases

> Collections themselves are grouped into **databases**.

Reviewer: ```suggestion
Collections are grouped into **databases**.
```
